Global anonymous functions name collisions

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.2.4 227d12f
**OS:** macOS 10.15

**What did you do?**

```
$ cat g.v
__global main = fn () { }
$
```

**What did you expect to see?**

That's probably an invalid name for a global and should fail at v compile time
 
**What did you see instead?**

c error:

```
/tmp/v/g.13203318637238796548.tmp.c:1743:1: warning: variable named 'main' with external linkage has undefined behavior [-Wmain]
/tmp/v/g.13203318637238796548.tmp.c:10608:5: error: redefinition of 'main' as different kind of symbol
int main(int ___argc, char** ___argv){
    ^
/tmp/v/g.13203318637238796548.tmp.c:1743:30: note: previous definition is here
anon_fn_d41a6318fa00fc6a__22 main;
                             ^
1 warning and 2 errors generated.
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2097,6 +2097,10 @@ fn (mut c Checker) global_decl(mut node ast.GlobalDecl) {
 			c.error('unknown type `$sym.name`', field.typ_pos)
 		}
 		if field.has_expr {
+			if field.expr is ast.AnonFn && field.name == 'main' {
+				c.error('the `main` function is the program entry point, cannot redefine it',
+					field.pos)
+			}
 			field.typ = c.expr(field.expr)
 			mut v := c.file.global_scope.find_global(field.name) or {
 				panic('internal compiler error - could not find global in scope')

--- a/vlib/v/checker/tests/globals/redefine_main.out
+++ b/vlib/v/checker/tests/globals/redefine_main.out
@@ -1,0 +1,3 @@
+vlib/v/checker/tests/globals/redefine_main.vv:1:10: error: the `main` function is the program entry point, cannot redefine it
+    1 | __global main = fn () int { return 22 }
+      |          ~~~~

--- a/vlib/v/checker/tests/globals/redefine_main.vv
+++ b/vlib/v/checker/tests/globals/redefine_main.vv
@@ -1,0 +1,1 @@
+__global main = fn () int { return 22 }

